### PR TITLE
fetchgithub: support for sparse checkouts

### PR DESCRIPTION
**Pull request by SCOTT-HAMILTON**
###### Motivation for this change
If we one day try to use some openbsd utilities as asked here https://github.com/NixOS/nixpkgs/issues/113483, we
will have either to build the entire openbsd project in a single derivation (not optimal) or to take only a few projects inside it and build them on their own.

This PR adds a feature to fetchFromGitHub to sparseCheckout a repo. This is for example what you would use to only extract openbsd's make implementation.
```nix
src = fetchFromGitHub {
  owner = "openbsd";
  repo = "src";
  rev = "e5d32326ca316ca97670d484b46db42e9d6da269";
  sparseCheckout = "usr.bin/make";
  sha256 = "12jicmj3liicp2j02jz0i4yan81m3i7w0gbfrqhm9bcckfhdh7jl";
};
```

This optimization reduces the band width usage which is already quite big when working on nixpkgs.

__________________

~~The implementation behind this feature uses [github's svn bridge](https://docs.github.com/en/github/importing-your-projects-to-github/support-for-subversion-clients) which seems to be the only way to fetch a github repo's sub/subdir from a specific commit.~~
 
# Warning

 - ~~We can't use the print_results mechanism that shows informations about the git fetch such has the full rev as it's obviously not compatible with svn. This could be fixed with a svn version though~~.
 - It should work with tags but other than that and commit shas, I might have missed some edge cases.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 113689` at e4a82e37 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/d309867aad5de2d178194cb28198214a">bundix</a></li></li>
    <li>cabal2nix</li>
    <li><a href="https://gist.github.com/178ee25053d319f43693840861b32794">crate2nix</a></li></li>
    <li><a href="https://gist.github.com/743f4955429ecbb49cbb45932559b4df">dep2nix</a></li></li>
    <li><a href="https://gist.github.com/37f8c81448f9ccc1110dd62182a2fe21">go2nix</a></li></li>
    <li><a href="https://gist.github.com/58451d2cdc54d23b575af0caaab80291">haskellPackages.update-nix-fetchgit</a></li></li>
    <li><a href="https://gist.github.com/9c80879e5197cc7ed956a3b8bfd72856">nix-prefetch-git</a></li></li>
    <li><a href="https://gist.github.com/9ae87cb54cac2fb9b11f8e9a61edeeff">nix-prefetch-scripts</a></li></li>
    <li><a href="https://gist.github.com/dc5aeb039515ee9ed1784a68cadc3c15">nix-update-source</a></li></li>
    <li><a href="https://gist.github.com/be4882b27582a9b1335bb2a384ab7f2d">update-nix-fetchgit</a></li></li>
    <li><a href="https://gist.github.com/00de8c22fc898f3e6628e484010d3e7c">vgo2nix</a></li></li>
  </ul>
</details>

**Comment by mohe2015** [comment]
@SCOTT-HAMILTON What about:
```bash
git clone --filter=tree:0 --no-checkout --depth=1 https://github.com/openbsd/src.git
cd src/
git sparse-checkout init --cone
git sparse-checkout set usr.bin/make
git fetch --depth=1 origin e5d32326ca316ca97670d484b46db42e9d6da269
git checkout e5d32326ca316ca97670d484b46db42e9d6da269
```

**Comment by SCOTT-HAMILTON** [comment]
This is good but it still fetches all the commits.

**Comment by sternenseemann** [comment]
I'd rather support `git sparse-checkout` in `fetchgit`, be it slower or not, than further increase the complexety of `fetchFromGitHub` with essentially a third way of fetching stuff from GitHub which uses a non-obvious method under non-obvious circumstances.

File: pkgs/build-support/fetchgithub/default.nix [hunk 1]
```diff
@@ -13,7 +14,7 @@ let
   useFetchGit = fetchSubmodules || (leaveDotGit == true) || deepClone;
```
Comment (sternenseemann) [inline pkgs/build-support/fetchgithub/default.nix:14]:
```suggestion
  useFetchGit = fetchSubmodules || (leaveDotGit == true) || deepClone || (sparseCheckout != null);
```

File: pkgs/build-support/fetchgithub/default.nix [hunk 2]
```diff
@@ -13,7 +14,7 @@ let
...
   # We prefer fetchzip in cases we don't need submodules as the hash
   # is more stable in that case.
-  fetcher = if useFetchGit then fetchgit else fetchzip;
+  fetcher = if useFetchGit || sparseCheckout != null then fetchgit else fetchzip;
```
Comment (sternenseemann) [inline pkgs/build-support/fetchgithub/default.nix:17]:
```suggestion
  fetcher = if useFetchGit then fetchgit else fetchzip;
```

File: pkgs/build-support/fetchgithub/default.nix [hunk 3]
```diff
@@ -28,9 +29,9 @@ let
     '';
     netrcImpureEnvVars = [ "${varBase}USERNAME" "${varBase}PASSWORD" ];
   };
-  fetcherArgs = (if useFetchGit
+  fetcherArgs = (if useFetchGit || sparseCheckout != null
```
Comment (sternenseemann) [inline pkgs/build-support/fetchgithub/default.nix:32]:
```suggestion
  fetcherArgs = (if useFetchGit
```

**Review by sternenseemann — CHANGES_REQUESTED** [review]

**Pull request closed without merging**